zipline/utils/data.py

```diff
@@ -391,3 +395,85 @@ def _update_buffer(self, frame):
+    function provided to the mapping.  Ties from the sort_key are broken by
+    comparing the original keys.  `iter` traverses the keys in sort order.
+
+    Parameters
+    ----------
+    mapping : mapping, optional
+    **kwargs
+        The initial mapping. This will double as the starting mapping so
+        the values must also be hashable.
+
+    >>> d = SortedDict(abs)
+    >>> d[-1] = 'negative one'
+    >>> d[0] = 'zero'
+    >>> d[2] = 'two'
+    >>> d  # doctest: +NORMALIZE_WHITESPACE
+    SortedDict[<built-in function abs>]([(0, 'zero'),
+                                         (-1, 'negative one'),
+                                         (2, 'two')])
+    >>> d[1] = 'one'  # Mutating the mapping maintains the sort order.
+    >>> d  # doctest: +NORMALIZE_WHITESPACE
+    SortedDict[<built-in function abs>]([(0, 'zero'),
+                                         (-1, 'negative one'),
+                                         (1, 'one'),
+                                         (2, 'two')])
+    >>> del d[0]
+    >>> d  # doctest: +NORMALIZE_WHITESPACE
+    SortedDict[<built-in function abs>]([(-1, 'negative one'),
+                                         (1, 'one'),
+                                         (2, 'two')])
+    >>> del d[2]
+    >>> d  # doctest: +NORMALIZE_WHITESPACE
+    SortedDict[<built-in function abs>]([(-1, 'negative one'),
+                                         (1, 'one')])
+    """
+    def __init__(self, sort_key, mapping=None, **kwargs):
+        self._map = {}
+        self._sorted_key_names = []
+        self._sort_key = sort_key
+
+        self.update(merge(mapping or {}, kwargs))
+
+    def __getitem__(self, name):
+        return self._map[name]
+
+    def __setitem__(self, name, value, _bisect_right=bisect.bisect_right):
+        self._map[name] = value
+        if len(self._map) > len(self._sorted_key_names):
+            key = self._sort_key(name)
+            pair = (key, name)
+            idx = _bisect_right(self._sorted_key_names, pair)
+            self._sorted_key_names.insert(idx, pair)
+
+    def __delitem__(self, name, _bisect_right=bisect.bisect_right):
+        del self._map[name]
+        idx = _bisect_right(self._sorted_key_names,
+                            (self._sort_key(name), name))
+        del self._sorted_key_names[idx - 1]
+
+    def __iter__(self):
+        for key, name in self._sorted_key_names:
+            yield name
+
+    def __len__(self):
+        return len(self._map)
+
+    def __repr__(self, _repr_running={}):
+        # Based on OrderedDict/defaultdict
+        call_key = id(self), get_ident()
+        if call_key in _repr_running:
+            return '...'
+        _repr_running[call_key] = 1
+        try:
+            if not self:
+                return '%s[%r]()' % (self.__class__.__name__, self._sort_key)
+            return '%s[%r](%r)' % (self.__class__.__name__, self._sort_key,
```

Comment:
also in py3 this only shows: `SortedDict[<function <lambda> at 0x7fae9a31cf28>](ItemsView(...))`, we might want `list(self.items())`